From: "Igor B. Poretsky" <poretsky@rakurs.com>
Date: Fri, 25 Nov 2022 14:57:05 +0300
Subject: Visual line reading enhancement

Distinguishable tones for various decorations.
More reliable auditory icon indication.
---
 lisp/emacspeak-speak.el | 46 +++++++++++++++++++++++++++++++---------------
 1 file changed, 31 insertions(+), 15 deletions(-)

diff --git a/lisp/emacspeak-speak.el b/lisp/emacspeak-speak.el
index 594827e..b7c60b0 100644
--- a/lisp/emacspeak-speak.el
+++ b/lisp/emacspeak-speak.el
@@ -1026,30 +1026,46 @@ Cues the start of a physical line with auditory icon `left'."
   (interactive)
   (cl-declare (special dtk-stop-immediately emacspeak-show-point))
   (when dtk-stop-immediately (dtk-stop))
-  (let ((inhibit-field-text-motion t)
-        (inhibit-read-only t)
-        (start nil)
-        (end nil)
-        (cursor-sensor-inhibit t)
-        (inhibit-modification-hooks t)
-        (deactivate-mark nil)
-        (line nil)
-        (orig (point)))
-    (cond
-     ((looking-at "^ *") (emacspeak-auditory-icon 'left))
-     ((looking-at " *$") (emacspeak-auditory-icon 'right)))
+  (let* ((inhibit-field-text-motion t)
+         (inhibit-read-only t)
+         (cursor-sensor-inhibit t)
+         (inhibit-modification-hooks t)
+         (deactivate-mark nil)
+         (start (line-beginning-position))
+         (end (line-end-position))
+         (line nil)
+         (orig (point)))
     (save-mark-and-excursion
       (beginning-of-visual-line)
-      (setq start (point))
+      (setq start (max (point) start))
       (end-of-visual-line)
-      (setq end (point))
+      (setq end (min (point) end))
+      (cond
+       ((and (= start (line-beginning-position)) (< end (line-end-position)))
+        (emacspeak-auditory-icon 'left))
+       ((and (> start (line-beginning-position)) (= end (line-end-position)))
+        (emacspeak-auditory-icon 'right)))
       (setq line
             (if emacspeak-show-point
                 (ems-set-personality-temporarily
                  orig (1+ orig)
                  voice-animate (buffer-substring start end))
               (buffer-substring start end)))
-      (dtk-speak line))))
+      (cond
+       ((string-equal ""  line)
+        (dtk-tone 130.8   75 'force))
+       ((string-match  emacspeak-speak-blank-line-regexp  line)
+        (dtk-tone 261.6   75 'force))
+       ((and (not (eq 'all dtk-punctuation-mode))
+             (string-match  emacspeak-horizontal-rule line))
+        (dtk-tone 523.3   100 t))
+       ((and (not (eq 'all dtk-punctuation-mode))
+             (string-match  emacspeak-decoration-rule line))
+        (dtk-tone 1047   100 t))
+       ((and (not (eq 'all dtk-punctuation-mode))
+             (string-match  emacspeak-unspeakable-rule line))
+        (dtk-tone 2093   100 t))
+       (t (dtk-speak line))))))
 
 (defvar-local emacspeak-speak-last-spoken-word-position nil
   "Records position of the last word that was spoken.
